rustls/src/server/tls12.rs

```diff
@@ -180,12 +180,35 @@ mod client_hello {
                 ));
             }
 
+            let groups_ext_or_default = groups_ext.unwrap_or(&[]);
             let group = self
                 .config
                 .provider
                 .kx_groups
                 .iter()
-                .find(|skxg| groups_ext.contains(&skxg.name()))
+                .filter(|skxg| skxg.name().key_exchange_algorithm() == Some(self.suite.kx))
+                .find(|skxg| groups_ext_or_default.contains(&skxg.name()))
+                .or_else(|| {
+                    // If kx for the selected cipher suite is DHE and no DHE groups are specified in the extension,
+                    // the server is free to choose DHE params, we choose the first DHE kx group of the provider.
+                    if self.suite.kx == KeyExchangeAlgorithm::DHE
+                        && !groups_ext_or_default
+                            .iter()
+                            .any(|g| g.key_exchange_algorithm() == Some(KeyExchangeAlgorithm::DHE))
+                    {
+                        trace!("No DHE groups specified in ClientHello groups extension, server choosing DHE parameters");
+                        self.config
+                            .provider
+                            .kx_groups
+                            .iter()
+                            .find(|skxg| {
+                                skxg.name().key_exchange_algorithm()
+                                    == Some(KeyExchangeAlgorithm::DHE)
+                            })
+                    } else {
+                        None
+                    }
+                })
```

Comment:
I think this combinator chain gets pretty unwieldy, I think it might be time to rewrite it into a `for`-loop?

Reply:
I'm not sure rewriting this into a for loop would improve it that much. Most of the complexity comes from `.or_else(..)` invocation, which I don't think we can get around easily.

Reply:
Hmm, maybe this should be a method on `ServerConfig` that takes `&ClientHelloPayload` and `KeyExchangeAlgorithm` and yields `Result<&dyn SupportedKxGroup, Error>`? What do you think?

Reply:
I tried simplifying it a bit first. Let me know what you think. If still too unwieldy, I can try your suggestion.

Reply:
I think I would still suggest changing this the way I proposed in my previous comment. It might even make sense to do that for the current setup in an early commit so that this commit can focus on the changes made for FFDHE.

Note in particular that I usually prefer not to have consecutive combinators like `.filter().find()`, since the closures can just be combined into one. I also expect that isolating this code in a separate function will bring together blocks of code that are currently separated (extracting `group_ext`, then finding the `group`) and will bring opportunities to "linearize" the code by using early returns to handle simple/exceptional cases.

Reply:
I added a `pick_kx_group()` function to `CompleteClientHelloHandling`. Let me know what you think. If it looks good, I'll squash dc5e74b15c1.

Reply:
I was suggesting that the method you've now called `pick_kx_group()` take `&ClientHelloPayload`, so we bring it together with the earlier block of code that defines `group_ext`. I think it would be good to update the `pick_kx_group()` commit to add that, and I'd actually prefer to keep that commit in the commit history, to keep the refactoring separate from the changes needed for FFDHE support.

Reply:
To do that, I'd need to take `Context` as well, so I can send alerts from `pick_kx_group()`. Should I do that?
Also note that strictly speaking, this would change the semantics of `handle_client_hello()`. E.g., currently, it checks for existence of the groups extension before checking resumption tickets. With the suggested change, we don't do that.

Reply:
If we get no complaints from bogo I think it's probably okay.